packages/vite/src/node/plugins/assetImportMetaUrl.ts

```diff
@@ -62,16 +65,31 @@ export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
           }
 
           const url = rawUrl.slice(1, -1)
-          const file = path.resolve(path.dirname(id), url)
-          // Get final asset URL. Catch error if the file does not exist,
-          // in which we can resort to the initial URL and let it resolve in runtime
-          const builtUrl = await fileToUrl(file, config, this).catch(() => {
+          assetResolver ??= config.createResolver({
+            root: config.publicDir,
+            extensions: [],
+            mainFields: [],
+            tryIndex: false,
+            preferRelative: true
+          })
```

Comment:
I think we should match the resolver used for `import.meta.glob('@/alias')` too?

https://github.com/vitejs/vite/blob/0d2089c6876bcc04a46c3f7947ce4effc91dd9fd/packages/vite/src/node/plugins/importAnalysis.ts#L111-L115

Not sure if that makes sense.

Re `root: config.publicDir`, if we follow above, this would be removed. If the user wants to use a public file, shouldn't they specify the string directly without `new URL`?

Reply:
For public files—they could use a relative path, though if they wanted to resolve it to an absolute URL, they'd use `new URL('path/to/file', self.location.href)` or `new URL('/baseDir/path/to/file', self.location.href)`. However, a `<base href>` tag can interfere with the former, and the latter requires that you embed the output base directory every time.

Regardless I think the existing `root` I have should probably be removed because it prevents other directories that might be at the project root from resolving—for example, the project root might have `img src static`, with code that uses `new URL('/img/logo.png', import.meta.url)`. The resolver would attempt to find the file at `/static/img/logo.png` and fail.

Off-topic, but why does `import.meta.glob` have `preferRelative: true`? This seems at odds with the Vite docs
> The glob patterns are treated like import specifiers: they must be either relative (start with `./`) or absolute (start with `/`, resolved relative to project root) or an alias path

Reply:
I believe `preferRelative: true` is so that `new URL("./foo.png", import.meta.url)` is the same as `new URL("foo.png", import.meta.url)` like you mentioned above. It's a very rare case and could be safe to remove too, but I think it's harmless to be there too.

I don't see how it's at odds with the docs though, but here's how `preferRelative` changes things in the resolver:

https://github.com/vitejs/vite/blob/3edd1af56e980aef56641a5a51cf2932bb580d41/packages/vite/src/node/plugins/resolve.ts#L170

It adds another check to see if it can be resolved relatively (though we would already have a fast-path for it).

Reply:
I pointed out that the existing glob behavior is at odds with the docs because right now, with `preferRelative: true`, relative imports do not have to start with `./`.

I just wanted to bring it to someone's attention in case it mattered. `preferRelative: true` definitely makes sense for these asset URLs here, though.

Reply:
I see, yeah that's true. We can look into this in the future to make it more restrictive. I think the implementation for `transformImportGlob` could be improved too to use a single resolver.